Core/Src/main.c

```diff
@@ -0,0 +1,935 @@
+static void MX_TIM1_Init(void);
+static void MX_TIM2_Init(void);
+static void MX_TIM16_Init(void);
+static void MX_TIM3_Init(void);
+static void MX_TIM4_Init(void);
+static void MX_TIM8_Init(void);
+static void MX_TIM6_Init(void);
+static void MX_TIM15_Init(void);
+/* USER CODE BEGIN PFP */
+void dshot600(uint32_t *motor, uint16_t value);
+uint16_t prepareDshotPacket(const uint16_t value);
+int16_t enc_val_prev[3];
+struct Data d1 = {1, 9, 3, 0};
+
+float required_speed = 4 ;
+float last_error, error, integral, derevative;
+float Kp=110, Ki=0.1, Kd=0.001;
+
+void HAL_TIM_PeriodElapsedCallback(TIM_HandleTypeDef *htim) {
+    if (htim == &htim1) {
+
+        uint16_t counts_now = ((TIM3->CNT)>>2);
+        uint16_t dc = counts_now - enc_val_prev[0];
+        enc_val_prev[0] = counts_now;
+
+        float rps = ((float )dc/24.f)*10;
+        d1.B=rps*100;
+
+
+
+        error = required_speed - rps;
+        integral = integral+error;
+        derevative = error-last_error;
+        last_error = error;
+        uint16_t control = (Kp * error) + (Ki * integral) + (Kd * derevative);
+        TIM1->CCR1 = control>100?100:control;
+
+        d1.C = control;
+
+        USBD_CUSTOM_HID_SendReport(&hUsbDeviceFS, (uint8_t *) &d1, sizeof(struct Data));
+
+        if (rps > 0.5f)
+            HAL_GPIO_TogglePin(GPIOB, GPIO_PIN_8);
+    }
+}
+/* USER CODE END PFP */
+
+/* Private user code ---------------------------------------------------------*/
+/* USER CODE BEGIN 0 */
+
+
+//extern volatile uint16_t DT;
+/* USER CODE END 0 */
+
+/**
+  * @brief  The application entry point.
+  * @retval int
+  */
+int main(void)
+{
+  /* USER CODE BEGIN 1 */
+
+  /* USER CODE END 1 */
+
+  /* MCU Configuration--------------------------------------------------------*/
+
+  /* Reset of all peripherals, Initializes the Flash interface and the Systick. */
+  HAL_Init();
+
+  /* USER CODE BEGIN Init */
+
+  /* USER CODE END Init */
+
+  /* Configure the system clock */
+  SystemClock_Config();
+
+  /* USER CODE BEGIN SysInit */
+
+  /* USER CODE END SysInit */
+
+  /* Initialize all configured peripherals */
+  MX_GPIO_Init();
+  MX_DMA_Init();
+  MX_TIM1_Init();
+  MX_USB_Device_Init();
+  MX_TIM2_Init();
+  MX_TIM16_Init();
+  MX_TIM3_Init();
+  MX_TIM4_Init();
+  MX_TIM8_Init();
+  MX_TIM6_Init();
+  MX_TIM15_Init();
+  /* USER CODE BEGIN 2 */
+    /*uint32_t esc_dshot_buffer[5] = {25, 50, 0, 25, 0};
+  */
+    //Motor motor = {&htim2, TIM_CHANNEL_1};
+    //dshot_init(DSHOT150, &motor);
+    //  htim2.Instance->CCR1 = 50;
+    //  HAL_TIM_PWM_Start(&htim2, TIM_CHANNEL_1);
+    //
+    //   HAL_Delay(5000);
+  /* USER CODE END 2 */
+
+  /* Infinite loop */
+  /* USER CODE BEGIN WHILE */
+
+    DWT_Delay_Init();
+
+
+    dshot600(motor, 0);
+    HAL_TIM_PWM_Start_DMA(&htim15, TIM_CHANNEL_1, motor, DSHOT_FRAME_SIZE);
+    HAL_Delay(2000);
+    dshot600(motor, 100);
+    HAL_TIM_PWM_Start_DMA(&htim15, TIM_CHANNEL_1, motor, DSHOT_FRAME_SIZE);
+    HAL_Delay(1000);

```

Comment:
`HAL_TIM_PWM_Start_DMA` could be called only once, since DMA is in circular mode.

To change the thrower speed, only the `motor` array would need be updated.